### ISSUE TITLE
Deadlock when remote function has `ray.shutdown`.

### ISSUE THREAD

**Issue by guoyuhong** [ray-project/ray#4495]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 
- **Ray installed from (source or binary)**: Source
- **Ray version**:
- **Python version**:
- **Exact command to reproduce**:

### Describe the problem

In `worker.py`, `self._process_task(task, execution_info)` is called in Function `_wait_for_and_process_task` within the worker lock. The remote call of `ray.shutdown` is waiting for the import thread to join and the import thread is also waiting for the worker lock at the same time. Therefore, `worker.import_thread.join_import_thread()` will never stop with this deadlock.

### Source code / logs

```
---------------one thread

  File "/usr/local/python3/lib/python3.6/threading.py", line 884, in _bootstrap
    self._bootstrap_inner()
  File "/usr/local/python3/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/python3/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Ray-package/ray/python/ray/import_thread.py", line 75, in _run
    with self.worker.lock:

---------------another thread

  File "/Ray-package/ray/python/ray/workers/default_worker.py", line 166, in <module>
    ray.worker.global_worker.main_loop()
  File "/Ray-package/ray/python/ray/worker.py", line 1030, in main_loop
    self._wait_for_and_process_task(task)
  File "/Ray-package/ray/python/ray/worker.py", line 981, in _wait_for_and_process_task
    self._process_task(task, execution_info)
  File "/Ray-package/ray/python/ray/worker.py", line 880, in _process_task
    self.actors[key], *arguments)
  File "/Ray-package/ray/python/ray/function_manager.py", line 785, in actor_method_executor
    method_returns = method(actor, *args)
  File "/home/admin/py3/lib/python3.6/site-packages/modelserving/model_serving.py", line 387, in stop
    ray.shutdown() # this will remove worker from local scheduler thus process will be free from ray
  File "/Ray-package/ray/python/ray/worker.py", line 1604, in shutdown
    disconnect()
  File "/Ray-package/ray/python/ray/worker.py", line 2218, in disconnect
    worker.import_thread.join_import_thread()
  File "/Ray-package/ray/python/ray/import_thread.py", line 46, in join_import_thread
    self.t.join()
  File "/usr/local/python3/lib/python3.6/threading.py", line 1056, in join
    self._wait_for_tstate_lock()
  File "/usr/local/python3/lib/python3.6/threading.py", line 1072, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <module>
```

**Comment by guoyuhong** [comment]
I will submit a fix later on.

**Comment by robertnishihara** [comment]
Thanks @guoyuhong. Are you saying there is some task like this?

```python
@ray.remote
def f():
    ray.shutdown()
```

I think `ray.shutdown` should only be called by the same process that called `ray.init()`, meaning the driver process and not the workers.

**Comment by guoyuhong** [comment]
@robertnishihara Yes, we have a scenario that we need to call `ray.shutdown()` to manually stop an actor to release the resources and start a new one. Actually, this `ray.shutdown()` is called corresponding to worker's `ray.init`.

**Comment by robertnishihara** [comment]
Only the driver calls `ray.init()`, right? Are you referring to the call to `connect()` in `default_worker.py`?

Why not just have the worker call `sys.exit()` or `__ray_terminate__()`?

**Comment by guoyuhong** [comment]
@robertnishihara Oh, in our repo there is bit different. We have an argument to control whether or not a `ray.shutdown` will clean the resource or not. Anyway, you way may still work.

I checked the code again. In driver, this deadlock may also happen, since the import thread starts in both worker and driver. Furthermore, I also think it is a good practice to separate the lock in worker and function manager instead of locking it everywhere. I have finished a PR to separate the thread lock.
